[RFC-17] Add WithdrawalRequestCanceled Event to DepositManager

Zena-park opened the pull request:
## Related RFC
- **RFC Discussion**: https://github.com/tokamak-network/tokamak-dao-contracts/discussions/17
- **Original Proposal**: https://github.com/tokamak-network/tokamak-dao-contracts/discussions/16

## Summary

This PR implements RFC-17 by adding a `WithdrawalRequestCanceled` event to the `DepositManager` contract. This enables accurate tracking of TON circulating supply by distinguishing between "fresh deposits" and "withdrawal cancellations (redeposits)".

## Changes

### Contract Changes
- ✅ Created `DepositManagerV1_2` contract with `WithdrawalRequestCanceled` event
- ✅ Modified `_redeposit` function to emit both `Deposited` and `WithdrawalRequestCanceled` events
- ✅ No breaking changes - only additive event emission

### Test Coverage
- ✅ Standalone tests: [`test/deposit-manager-v1-2-standalone.test.ts`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/test/deposit-manager-v1-2-standalone.test.ts)
- ✅ Agenda-based registration tests: [`test/deposit-manager-v1-2-agenda.test.ts`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/test/deposit-manager-v1-2-agenda.test.ts)
- ✅ Shared helper functions: [`test/shared/depositManagerHelpers.ts`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/test/shared/depositManagerHelpers.ts)
- ✅ All tests passing

## Implementation Details

When `redeposit` or `redepositMulti` is called, both events are emitted:
```solidity
emit Deposited(layer2, msg.sender, accAmount);
emit WithdrawalRequestCanceled(layer2, msg.sender, accAmount);
```

## Testing

All tests passing ✅
- Event emission verification
- E2E flow testing
- Agenda-based registration testing
- Edge cases (rounding, multiple withdrawals)

<img width="1004" height="753" alt="스크린샷 2025-12-05 오후 7 27 11" src="https://github.com/user-attachments/assets/15127f42-dbd7-4950-a02b-34fac44cdff0" />
<img width="937" height="859" alt="스크린샷 2025-12-05 오후 7 27 52" src="https://github.com/user-attachments/assets/c1971c9c-f117-4209-93c5-357f812cb2ce" />

## Impact
- **Breaking Changes**: ❌ None 
- **Security**: No security impact (event-only change)

## References
- RFC document: [`docs/RFC/RFC-17-DepositManagerV1_2.md`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/docs/RFC/RFC-17-DepositManagerV1_2.md) 
- Contract: [`contracts/stake/managers/DepositManagerV1_2.sol`](https://github.com/tokamak-network/ton-staking-v2/blob/rfc-17/contracts/stake/managers/DepositManagerV1_2.sol)

Zena-park committed:
feat: Add WithdrawalRequestCanceled event to DepositManagerV1_2

- Add DepositManagerV1_2 contract with WithdrawalRequestCanceled event
- Update DepositManager_Dune.json ABI with new event and functions
- Add comprehensive unit and E2E tests for redeposit functionality
- Implement helper functions for test modularity
- Use 27 decimals for staking balance comparisons

Zena-park committed:
Add DepositManagerV1_2 tests with agenda-based registration

- Add deposit-manager-v1-2-standalone.test.ts: Standalone tests for DepositManagerV1_2
  - Tests WithdrawalRequestCanceled event emission
  - Tests both Deposited and WithdrawalRequestCanceled events for redeposit/redepositMulti
  - Tests E2E flow: Deposit -> RequestWithdrawal -> Redeposit

- Add deposit-manager-v1-2-agenda.test.ts: Agenda-based registration tests
  - Tests full DAO agenda process: creation -> voting -> execution
  - Verifies DepositManagerV1_2 registration via agenda
  - Tests functionality after agenda execution

- Add shared/depositManagerHelpers.ts: Common helper functions
  - Refactor common test utilities (setupDeposit, setupWithdrawal, etc.)
  - Shared by both test files for better maintainability
  - Includes STAKING_DECIMALS constant and helper functions

Zena-park committed:
Update RFC-17 documentation and contract references

- Add Reference Implementation section with code snippets
- Update contract comments with RFC and proposal links
- Add Security Considerations section
- Update References section with correct discussion links

Zena-park committed:
docs: Update RFC-17 document with repository references and remove old RFC file

Zena-park committed:
fix: resolve merge conflict and apply code review feedback

- Resolve merge conflict in RFC-17-DepositManagerV1_2.md
- Fix error message: 'Caller is not a Layer2' -> 'Address is not a registered Layer2'
- Fix typo: requsts -> requests (local variable name)

Zena-park commented on contracts/stake/managers/DepositManagerV1_2.sol:
```diff
@@ -0,0 +1,114 @@
+    function _redeposit(address layer2, uint256 i, uint256 n) internal onlyLayer2(layer2) returns (bool) {
+      uint256 accAmount;
+
+      WithdrawalReqeust[] memory requsts = _withdrawalRequests[layer2][msg.sender];
```
Thank you! 
I've fixed the typo in the local variable: `requsts` → `requests`.

The `WithdrawalReqeust` type name is part of the storage layout, so I've left it unchanged to avoid compatibility issues. I'll fix it when we have a storage layout change in the future.

Fix commit: [5aa5eaf8](https://github.com/tokamak-network/ton-staking-v2/commit/5aa5eaf8)

Zena-park commented on contracts/stake/managers/DepositManagerV1_2.sol:
```diff
@@ -0,0 +1,114 @@
+    ////////////////////
+
+    modifier onlyLayer2(address layer2) {
+      require(ILayer2Registry(_registry).layer2s(layer2), "Caller is not a Layer2");
```
Thank you! I've updated the error message to "Address is not a registered Layer2".

Fix commit: [5aa5eaf8](https://github.com/tokamak-network/ton-staking-v2/commit/5aa5eaf8)

Zena-park committed:
fix: optimize _redeposit function for gas efficiency

- Replace array memory copy with length-only read
- Use storage pointer instead of memory copy and write-back
- Remove unnecessary storage write inside loop
- Addresses gas optimization feedback from PR review

Fixes: #330 discussion_r2592152704

Zena-park commented on contracts/stake/managers/DepositManagerV1_2.sol:
```diff
@@ -0,0 +1,114 @@
+        accAmount = accAmount + amount;
+        r.processed = true;
+        _withdrawalRequests[layer2][msg.sender][i] = r;
+      }
```
Thank you! I've applied the gas optimization improvements.

- Removed array memory copy → read only length
- Used storage pointer → direct modification
- Removed storage write in loop

**Commit**: [21d21c9](https://github.com/tokamak-network/ton-staking-v2/pull/330/commits/21d21c96aef082c603d11dae1277e4db31229a37)

The array growth issue will be considered in future architectural improvements.

Zena-park committed:
docs: Update RFC-17 with PR link

shinthom commented on contracts/stake/managers/DepositManagerV1_2.sol:
```diff
@@ -0,0 +1,109 @@
+      emit Deposited(layer2, msg.sender, accAmount);
+
+      // add event for withdrawal request canceled
+      emit WithdrawalRequestCanceled(layer2, msg.sender, accAmount);
```
I’d like to suggest a small improvement for better readability:
- Place each event (emit) after its corresponding state variables are updated.
- Since the logic is essentially “cancel and then re-deposit,” it may be clearer to update the order accordingly as well (purely for readability).

For example:

```solidity
_withdrawalRequestIndex[layer2][msg.sender] += n;

// withdrawal-related storages
_pendingUnstaked[layer2][msg.sender] = _pendingUnstaked[layer2][msg.sender] - accAmount;
_pendingUnstakedLayer2[layer2] = _pendingUnstakedLayer2[layer2] - accAmount;
_pendingUnstakedAccount[msg.sender] = _pendingUnstakedAccount[msg.sender] - accAmount;
emit WithdrawalRequestCanceled(layer2, msg.sender, accAmount);

// deposit-related storages
_accStaked[layer2][msg.sender] = _accStaked[layer2][msg.sender] + accAmount;
_accStakedLayer2[layer2] = _accStakedLayer2[layer2] + accAmount;
_accStakedAccount[msg.sender] = _accStakedAccount[msg.sender] + accAmount;
emit Deposited(layer2, msg.sender, accAmount);
```

Zena-park replied:
Thank you! I've reordered the code as suggested.

**Applied commit**: [4732e9c3](https://github.com/tokamak-network/ton-staking-v2/commit/4732e9c3)

Zena-park replied:
@shinthom Does this change look good to you? If everything is clear, may I resolve this conversation?

shinthom replied:
Sure, good to me!

Zena-park committed:
refactor: improve code readability by reordering state updates and events

- Reorder logic to follow 'cancel then re-deposit' flow
- Emit WithdrawalRequestCanceled event immediately after withdrawal state updates
- Emit Deposited event immediately after deposit state updates
- Improves code readability and event log clarity

Addresses PR #330 review feedback from @shinthom

zzooppii commented:
Are you testing the version currently being tested on mainnet?

When creating an agenda, the method of creating an agenda differs depending on whether it is created before Agenda15 is passed or not, so please check only that part.

Zena-park commented:
> Are you testing the version currently being tested on mainnet?
> 
> When creating an agenda, the method of creating an agenda differs depending on whether it is created before Agenda15 is passed or not, so please check only that part.

@zzooppii I haven't simulated the actual agenda creation script yet.
I will write the simulation script first and let you know once it's ready. I would appreciate it if you could review it again at that time.
